### ISSUE TITLE
🚀 Feature: Add support for Cloudflare R2 Storage

### ISSUE THREAD

**Issue by stnguyen90** [appwrite/appwrite#3352]
### 🔖 Feature description

[Cloudflare R2 Storage](https://www.cloudflare.com/products/r2/) is an S3-compatible distributed object storage with generous [pricing and a free tier](https://developers.cloudflare.com/r2/platform/pricing/). It would be great if Appwrite could make use of this storage provider.

### 🎤 Pitch

A storage provider with a free tier would make switching to and testing a cloud storage provider easier.

Related issue:

* https://github.com/utopia-php/storage/issues/52

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by gewenyu99** [comment]
If it's S3 compatible, can it be supported rn by using the configurable ENV Vars? (I'm genuinely curious as a sanity check).

**Comment by stnguyen90** [comment]
> If it's S3 compatible, can it be supported rn by using the configurable ENV Vars? (I'm genuinely curious as a sanity check).

The way utopia storage works right now, a new Device needs to be implemented to change some parameters:
1. [Backblaze](https://github.com/utopia-php/storage/blob/83bac7127b038b5c09cdce42e397db5931b0f34f/src/Storage/Device/Backblaze.php#L35)
1. [DOSpaces](https://github.com/utopia-php/storage/blob/83bac7127b038b5c09cdce42e397db5931b0f34f/src/Storage/Device/DOSpaces.php#L33)
1. [Linode](https://github.com/utopia-php/storage/blob/83bac7127b038b5c09cdce42e397db5931b0f34f/src/Storage/Device/Linode.php#L32)
1. [Wasabi](https://github.com/utopia-php/storage/blob/83bac7127b038b5c09cdce42e397db5931b0f34f/src/Storage/Device/Wasabi.php#L38) 

The region constants for each provider are also a little nice to make sure you're not using the wrong region.

Then, the [Appwrite code](https://github.com/appwrite/appwrite/blob/fb8c5f6140f5f4d98351b88578d7daf84ecd0a32/app/init.php#L904-L946) needs to be updated to use the right device based on the `_APP_STORAGE_DEVICE` env var.

That said, there is an issue for creating a [generic S3 compatible device](https://github.com/utopia-php/storage/issues/28). And, I've seen a PR for a [generic S3](https://github.com/utopia-php/storage/pull/45), but it doesn't look like it implements 1 S3 device that you can use for all S3 compatible providers.

**Comment by gewenyu99** [comment]
Cool, I always learn a lot from talking to you :P

**Comment by Meldiron** [comment]
@gewenyu99 I think we had PR to do that at some point before 0.14. Not sure why it was not used.

**Comment by runesign** [comment]
I have the same wish!

**Comment by doublevcodes** [comment]
Hello, I'd like to have a go at implementing this!

**Comment by stnguyen90** [comment]
@doublevcodes, thanks for your interest! 🙏 Happy hacking!

**Comment by doublevcodes** [comment]
So as I understood at it, I need to create 2 PRs? One to utopia-php/storage and one to this repo

**Comment by Meldiron** [comment]
@doublevcodes That is correct 😇 You can take a look at this Hacktoberfest issue that we created, it's on the same topic but different adapter. You can use it's description to learn more: https://github.com/appwrite/appwrite/issues/3991

**Comment by doublevcodes** [comment]
Hi, I'm a bit confused on how to implement this because of how Cloudflare R2 bucket URLs are structured.

So far I have something like this:

```php
  public function __construct(string $root, string $accountId, string $accessKey, string $secretKey, string $bucket, string $region = self::AUTO, string $acl = self::ACL_PRIVATE)
  {
      parent::__construct($root, $accessKey, $secretKey, $bucket, $region, $acl);
      $this->headers['host'] = $accountId . '.' . 'r2' . '.' . 'cloudflarestorage' . '.' . 'com' . '/'. $bucket;
  }
```

However I'm sure that this wouldn't form a valid `Host` header due to the slash.
How would I go about this?

**Comment by stnguyen90** [comment]
@doublevcodes, oye that's unfortunate Cloudflare does it differently...It looks like the implementation for Cloudflare won't be the same as the other S3 compatible providers like Linode. You'll need to figure out what else needs to be overridden in the base [S3](https://github.com/utopia-php/storage/blob/0.11.0/src/Storage/Device/S3.php) class and implement it in the Cloudflare class.

**Comment by doublevcodes** [comment]
Ah. Thanks, I'll have a look at what I can do.

**Comment by stnguyen90** [comment]
@doublevcodes, FYI, [virtual-hosted style paths](https://developers.cloudflare.com/r2/platform/changelog/#2022-05-16) should also be supported in Cloudflare

**Comment by pietrodicaprio** [comment]
> Hi, I'm a bit confused on how to implement this because of how Cloudflare R2 bucket URLs are structured.
> 
> So far I have something like this:
> 
> ```
>   public function __construct(string $root, string $accountId, string $accessKey, string $secretKey, string $bucket, string $region = self::AUTO, string $acl = self::ACL_PRIVATE)
>   {
>       parent::__construct($root, $accessKey, $secretKey, $bucket, $region, $acl);
>       $this->headers['host'] = $accountId . '.' . 'r2' . '.' . 'cloudflarestorage' . '.' . 'com' . '/'. $bucket;
>   }
> ```
> 
> However I'm sure that this wouldn't form a valid `Host` header due to the slash. How would I go about this?

Hello there, I was looking around for R2 support... You are not forced to keep the `bucket` at the end of the URL. You should use the URL without the bucket name and then use the name as usually done with standard S3 implementation.\
Here a python example using the official boto3 library for AWS:
having as URL provided by Cloudflare `https://blahblahblahblahblahblah.r2.cloudflarestorage.com/bucket-name-here`
```python
s3 = boto3.client('s3',
                      aws_access_key_id=key,
                      aws_secret_access_key=secret,
                      endpoint_url="https://blahblahblahblahblahblah.r2.cloudflarestorage.com")
 with open("some_file_name.png", "rb") as f:
        s3.upload_fileobj(f, "bucket-name-here", "some_file_name.png")
```
hope this helps a bit

**Comment by eldadfux** [comment]
Thank you everyone for celebrating Hacktoberfest 22 with us! This issue will now be closed as we're getting ready to celebrate Hacktoberfest 23.

**Comment by pietrodicaprio** [comment]
Dear @eldadfux I'm not sure this issue is related to an Hacktoberfest. It is a general request for the product's improvement

**Comment by gewenyu99** [comment]
> Dear @eldadfux I'm not sure this issue is related to an Hacktoberfest. It is a general request for the product's improvement

Our team opened this task for Hacktoberfest :P @stnguyen90 is a part of our team. AFAIK, he doesn't actually want this feature :D

**Comment by pietrodicaprio** [comment]
> > Dear @eldadfux I'm not sure this issue is related to an Hacktoberfest. It is a general request for the product's improvement
> 
> Our team opened this task for Hacktoberfest :P @stnguyen90 is a part of our team. AFAIK, he doesn't actually want this feature :D

Well, https://github.com/utopia-php/storage/pull/76 is a thing and @stnguyen90 requested a review 3 weeks ago so it looks like to be almost ready to be merged?

**Comment by gewenyu99** [comment]
> > > Dear @eldadfux I'm not sure this issue is related to an Hacktoberfest. It is a general request for the product's improvement
> > 
> > 
> > Our team opened this task for Hacktoberfest :P @stnguyen90 is a part of our team. AFAIK, he doesn't actually want this feature :D
> 
> Well, [utopia-php/storage#76](https://github.com/utopia-php/storage/pull/76) is a thing and @stnguyen90 requested a review 3 weeks ago so it looks like to be almost ready to be merged?

Yep, that will be a generic adaptor, too. So this will be supported in some future version along with all S3 compatible storage services 🤞

**Comment by AndrewBucklin** [comment]
Could we re-open this? Cloudflare R2 provides much better costs than AWS and it's S3 compliant. It already use it in other projects by simply entering the Cloudflare R2 storage details where the AWS S3 information is usually entered.

**Comment by JosefJezek** [comment]
please re-open this

**Comment by Tushar98644** [comment]
@stnguyen90 Hey , can i take up the issue ?

**Comment by Haimantika** [comment]
> @stnguyen90 Hey , can i take up the issue ?

Assigning it to you. Please remember to update us every 3 days and if you have any questions, do not hesitate to reach us out on [Discord](https://appwrite.io/discord)

**Comment by Tushar98644** [comment]
@stnguyen90 @gewenyu99 any updates on my pr?

**Comment by tessamero** [comment]
@Tushar98644 the PR is still awaiting your follow up before we can merge

**Comment by docimin** [comment]
any update? I think those 3 days are kind of over

**Comment by docimin** [comment]
@tessamero Is this implemented yet?

**Comment by tessamero** [comment]
@schneidermr , reopened the issue, thank you for submitting a PR :)

**Comment by stnguyen90** [comment]
@pietrodicaprio, thanks for the insight. Does this work too?

```
https://bucket-name-here.blahblahblahblahblahblah.r2.cloudflarestorage.com/
```

**Comment by docimin** [comment]
@stnguyen90 It does not.
It has to be: https://accountID.r2.cloudflarestorage.com/bucketname

**Comment by stnguyen90** [comment]
> @stnguyen90 It does not. It has to be: https://accountID.r2.cloudflarestorage.com/bucketname or https://accountID.eu.r2.cloudflarestorage.com/bucketname if you use EU juristiction

I forgot I actually researched this..According to [this](https://developers.cloudflare.com/r2/reference/changelog/#2022-05-16), it is supported.

**Comment by eweren** [comment]
Is there still anything that blocks [the related PR](https://github.com/appwrite/appwrite/pull/7172) from being merged? R2 support would be awesome 🚀

**Comment by docimin** [comment]
> Is there still anything that blocks [the related PR](https://github.com/appwrite/appwrite/pull/7172) from being merged? R2 support would be awesome 🚀

Heyo,
This PR also still needs to be merged, but I think the waiting currently is on @schneidermr 
https://github.com/utopia-php/storage/pull/103

**Comment by schneidermr** [comment]
> > Is there still anything that blocks [the related PR](https://github.com/appwrite/appwrite/pull/7172) from being merged? R2 support would be awesome 🚀
> 
> Heyo, This PR also still needs to be merged, but I think the waiting currently is on @schneidermr [utopia-php/storage#103](https://github.com/utopia-php/storage/pull/103)

They will let me know if that's the case, but if you could understand the result of the unit test, you would see that there is also some problem with the local storage adapter test case. So my changes absolutely couldn't break any unit test with that extra function param what is optional and not required. I'm happy that you need this changes, but in the future please don't write these type of comments if you aren't senior enough to understand the changes in the code and the result of the unit tests. 🙂

**Comment by docimin** [comment]
> > > Is there still anything that blocks [the related PR](https://github.com/appwrite/appwrite/pull/7172) from being merged? R2 support would be awesome 🚀
> 
> > 
> 
> > Heyo, This PR also still needs to be merged, but I think the waiting currently is on @schneidermr [utopia-php/storage#103](https://github.com/utopia-php/storage/pull/103)
> 
> 
> 
> They will let me know if that's the case, but if you could understand the result of the unit test, you would see that there is also some problem with the local storage adapter test case. So my changes absolutely couldn't break any unit test with that extra function param what is optional and not required. I'm happy that you need this changes, but in the future please don't write these type of comments if you aren't senior enough to understand the changes in the code and the result of the unit tests. 🙂

No need to be passive aggressive 😕. I'm just trying to help. The answer is based on the last comment on there.

I also wrote without assurance, which you most likely didn't read based on your answer. Just because you got mentioned doesn't automatically mean everyone is waiting for you and it's your fault. 😊

**Comment by adityaoberai** [comment]
Closing this issue. Thank you so much for participating in Hacktoberfest 2023! We can't wait to welcome you all during HF 2024! Stay tuned for a lot more amazing issues from the Appwrite team!

**Comment by pietrodicaprio** [comment]
Isn't https://github.com/utopia-php/storage/pull/103 solving https://github.com/utopia-php/storage/issues/28 that allows any S3-like storage solving also the mentioned  https://github.com/utopia-php/storage/pull/76 ?
With that merge it should be possible to use R2

**Comment by docimin** [comment]
> Isn't [utopia-php/storage#103](https://github.com/utopia-php/storage/pull/103) solving [utopia-php/storage#28](https://github.com/utopia-php/storage/issues/28) that allows any S3-like storage solving also the mentioned [utopia-php/storage#76](https://github.com/utopia-php/storage/pull/76) ? With that merge it should be possible to use R2

76 and 28 will not be merged, as 103 already does this I think, however due to test issues, https://github.com/utopia-php/storage/pull/106 got created. So I would say keep checking 106.

**Comment by eweren** [comment]
Hey there 👋 
Any updates? The related PR utopia-php/storage#106 seems to have been merged.

**Comment by zengxs** [comment]
Hello,

It seems that the utopia-php/storage now supports custom S3 endpoints, and the feature has been released. To enable support for Cloudflare R2 Storage in Appwrite, I believe we just need to add a configuration option for custom S3 endpoints.

There is a related PR (https://github.com/appwrite/appwrite/pull/7172) that addresses this by adding support for custom endpoint URLs for S3. Could this PR be reviewed and merged soon to enable this functionality?

Thank you for your time and attention!
